File: src/main/java/chapter7/DoOnExample.java
package chapter7;

import common.CommonUtils;
import common.Log;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

import java.util.concurrent.TimeUnit;


public class DoOnExample {

    public void basic() {
        String[] objs = {"1", "3", "5"};
        Observable<String> source = Observable.fromArray(objs);

        source.doOnNext(data -> Log.it("onNext / " + data))
                .doOnComplete(() -> Log.it("onComplete"))
                .doOnError(e -> Log.it("onError / " + e.getMessage()))
                .subscribe(Log::it);
    }

    public void withError() {
        Integer[] divider = {10, 5, 0};

        Observable.fromArray(divider)
                .map(div -> 1000 / div)
                .doOnNext(data -> Log.it("onNext / " + data))
                .doOnComplete(() -> Log.it("onComplete"))
                .doOnError(e -> Log.it("onError / " + e.getMessage()))
                .subscribe(Log::it);
    }

    public void doOnEach() {
        String[] data = {"ONE", "TWO", "THREE"};
        Observable<String> source = Observable.fromArray(data);

        source.doOnEach(noti -> {
            if (noti.isOnNext()) {
                Log.it("onNext / " + noti.getValue());
            } else if (noti.isOnComplete()) {
                Log.it("onComplete");
            } else if (noti.isOnError()) {
                Log.it("onError / " + noti.getError().getMessage());
            }
        }).subscribe(Log::it);
    }

    public void doOnEachObserver() {
        String[] objs = {"1", "3", "5"};
        Observable<String> source = Observable.fromArray(objs);

        source.doOnEach(new Observer<String>() {
            @Override
            public void onSubscribe(Disposable d) {
                // doOnEach()에서는 onSubscribe() 함수가 호출되지 않습니다.
            }

            @Override
            public void onNext(String s) {
                Log.it("onNext / " + s);
            }

            @Override
            public void onError(Throwable e) {
                Log.it("onError / " + e.getMessage());
            }

            @Override
            public void onComplete() {
                Log.it("onComplete");
            }
        }).subscribe(Log::it);
    }

    public void doOnSubscribeAndDispose() {
        String[] objs = {"1", "3", "5", "2", "6"};
        Observable<String> source = Observable.fromArray(objs)
                .zipWith(Observable.interval(100L, TimeUnit.MILLISECONDS), (a, b) -> a)
                .doOnSubscribe(notUsed -> Log.it("onSubscribe"))
                .doOnDispose(() -> Log.it("onDispose"));
        Disposable d = source.subscribe(Log::it);

        CommonUtils.sleep(200);
        d.dispose();
        CommonUtils.sleep(300);
    }

    public void doOnLifeCycle() {
        String[] objs = {"1", "3", "5", "2", "6"};
        Observable<String> source = Observable.fromArray(objs)
                .zipWith(Observable.interval(100L, TimeUnit.MILLISECONDS), (a, b) -> a)
                .doOnLifecycle(
                        d -> Log.it("onSubscribe"), () -> Log.it("onDispose"));
        Disposable d = source.subscribe(Log::it);

        CommonUtils.sleep(200);
        d.dispose();
        CommonUtils.sleep(300);
    }

    public void doOnTerminate() {
        String[] objs = {"1", "3", "5"};
        Observable<String> source = Observable.fromArray(objs);

        source.doOnTerminate(() -> Log.it("onTerminate"))
                .doOnComplete(() -> Log.it("onComplete"))
                .doOnError(e -> Log.it("onError / " + e.getMessage()))
                .subscribe(Log::it);
    }

    public static void main(String[] args) {
        DoOnExample demo = new DoOnExample();
        // demo.basic();
        // demo.withError();
        // demo.doOnEach();
        // demo.doOnEachObserver();
        // demo.doOnSubscribeAndDispose();
        // demo.doOnLifeCycle();
        demo.doOnTerminate();
    }

}

File: src/main/kotlin/chapter7/DoOnExample_.kt
package chapter7

import common.CommonUtils_
import common.Log_
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import io.reactivex.functions.BiFunction
import java.util.concurrent.TimeUnit


class DoOnExample_ {

    fun basic() {
        val objs = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*objs)

        source.doOnNext { Log_.it("onNext / $it") }
                .doOnComplete { Log_.it("onComplete") }
                .doOnError { Log_.it("onError / ${it.message}") }
                .subscribe { Log_.it(it) }
    }

    fun withError() {
        val divider = arrayOf(10, 5, 0)

        Observable.fromArray(*divider)
                .map { 1000 / it }
                .doOnNext { Log_.it("onNext / $it") }
                .doOnComplete { Log_.it("onComplete") }
                .doOnError { Log_.it("onError / ${it.message}") }
                .subscribe { Log_.it(it) }
    }

    fun doOnEach() {
        val data = arrayOf("ONE", "TWO", "THREE")
        val source = Observable.fromArray(*data)

        source.doOnEach {
            when {
                it.isOnNext -> Log_.it("onNext / ${it.value}")
                it.isOnComplete -> Log_.it("onComplete")
                it.isOnError -> Log_.it("onError / ${it.error?.message}")
            }
        }.subscribe { Log_.it(it) }
    }

    fun doOnEachObserver() {
        val objs = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*objs)

        source.doOnEach(object : Observer<String> {
            override fun onSubscribe(d: Disposable) {
                // doOnEach()에서는 onSubscribe() 함수가 호출되지 않습니다.
            }

            override fun onNext(s: String) {
                Log_.it("onNext / $s")
            }

            override fun onError(e: Throwable) {
                Log_.it("onError / ${e.message}")
            }

            override fun onComplete() {
                Log_.it("onComplete")
            }
        }).subscribe { Log_.it(it) }
    }

    fun doOnSubscribeAndDispose() {
        val objs = arrayOf("1", "3", "5", "2", "6")
        val source = Observable.fromArray(*objs)
                .zipWith(Observable.interval(100L, TimeUnit.MILLISECONDS), BiFunction { t1: String, _: Long -> t1 })
                .doOnSubscribe { Log_.it("onSubscribe") }
                .doOnDispose { Log_.it("onDispose") }
        val d = source.subscribe { Log_.it(it) }

        CommonUtils_.sleep(200)
        d.dispose()
        CommonUtils_.sleep(300)
    }

    fun doOnLifeCycle() {
        val objs = arrayOf("1", "3", "5", "2", "6")
        val source = Observable.fromArray(*objs)
                .zipWith(Observable.interval(100L, TimeUnit.MILLISECONDS), BiFunction { t1: String, _: Long -> t1 })
                .doOnLifecycle(
                        { Log_.it("onSubscribe") },
                        { Log_.it("onDispose") }
                )
        val d = source.subscribe { Log_.it(it) }

        CommonUtils_.sleep(200)
        d.dispose()
        CommonUtils_.sleep(300)
    }

    fun doOnTerminate() {
        val objs = arrayOf("1", "3", "5")
        val source = Observable.fromArray(*objs)

        source.doOnTerminate { Log_.it("onTerminate") }
                .doOnComplete { Log_.it("onComplete") }
                .doOnError { Log_.it("onError / ${it.message}") }
                .subscribe { Log_.it(it) }
    }

}


fun main(args: Array<String>) {
    val demo = DoOnExample_()
    // demo.basic()
    // demo.withError()
    // demo.doOnEach()
    // demo.doOnEachObserver()
    // demo.doOnSubscribeAndDispose()
    // demo.doOnLifeCycle()
    demo.doOnTerminate()
}
